torch/csrc/jit/script/compiler.cpp

```diff
@@ -1396,17 +1397,69 @@ struct to_ir {
-    // TODO: start, stop, step loop
-    if (args.size() != 1) {
+    Value *end_val, *start_val, *step_val;
+    bool isSimpleRange = (args.size() == 1);
+    if (isSimpleRange) {
+      end_val = ensureInt(range, emitExpr(args[0]));
+    } else if (args.size() == 2) {
+      start_val = ensureInt(range, emitExpr(args[0]));
+      end_val = ensureInt(range, emitExpr(args[1]));
+      step_val = end_val->owningGraph()->insertConstant(1);
+    } else if (args.size() == 3) {
+      start_val = ensureInt(range, emitExpr(args[0]));
+      end_val = ensureInt(range, emitExpr(args[1]));
+      step_val = ensureInt(range, emitExpr(args[2]));
+    } else {
       throw ErrorReport(range)
-          << "range() expects 1 argument but got " << args.size();
+          << "range() expects 1-3 arguments but got " << args.size();
     }
-    auto max_trip_count_val = ensureInt(range, emitExpr(args[0]));
     const auto& ident_name = target.name();
-    auto assigner = [ident_name, range](
+    auto assigner = [ident_name, range, start_val, step_val, isSimpleRange](
                         Value* index, std::shared_ptr<Environment> env) {
-      env->setVar(range, ident_name, index);
+      Value* derived_index;
+      if (isSimpleRange) {
+        derived_index = index;
+      } else {
+        auto g = index->owningGraph();
+        auto offset = g->insertNode(g->create(aten::mul, {index, step_val}, 1))
+                          ->output()
+                          ->setType(IntType::get());
+        derived_index =
+            g->insertNode(g->create(aten::add, {offset, start_val}, 1))
+                ->output()
+                ->setType(IntType::get());
+      }
+      env->setVar(range, ident_name, derived_index);
     };
+    Value* max_trip_count_val;
+    if (isSimpleRange) {
+      max_trip_count_val = end_val;
+    } else {
+      auto g = start_val->owningGraph();
+      // (abs(end-start) + abs(step)-1)/abs(step) = # of iterations
+      auto diff = g->insertNode(g->create(aten::sub, {end_val, start_val}, 1))
+                      ->output()
+                      ->setType(IntType::get());
+      auto diff_abs = g->insertNode(g->create(prim::abs, {diff}, 1))
+                          ->output()
+                          ->setType(IntType::get());
+      auto step_abs = g->insertNode(g->create(prim::abs, {step_val}, 1))
+                          ->output()
+                          ->setType(IntType::get());
+      auto ceil_offset =
+          g->insertNode(
+               g->create(aten::sub, {step_abs, g->insertConstant(1)}, 1))
+              ->output()
+              ->setType(IntType::get());
+      auto diff_sum =
+          g->insertNode(g->create(aten::add, {diff_abs, ceil_offset}, 1))
+              ->output()
+              ->setType(IntType::get());
+      max_trip_count_val =
+          g->insertNode(g->create(aten::floordiv, {diff_sum, step_abs}, 1))
+              ->output()
+              ->setType(IntType::get());
```

Comment:
It might be worth just copy-pasting the [python implementation](https://github.com/python/cpython/blob/22c526394b2ef51b985873ddbfbcc32c16411919/Objects/rangeobject.c#L838) of this into a `prim::` operator and just calling that here

Reply:
Isn't that a little bit too specific of an operator? I see this accomplishing 2 main benefits: 
1. It shortens the code significantly.
2. It speeds up the operation (I think?)

I don't think 2. matters much, and I think we can get much of the benefit by simply factoring out these operations to a function.

Reply:
This would be another acceptable alternative.

Reply:
My main motivation is actually to ensure that our calculation here is 100% consistent with Python. There really aren't any costs associated with us having a fine-grained operator. I think it's fine to have a private operator like `prim::__range_length` for internal use.

And on the perf point, our operator dispatch and operand handling are very heavyweight, so it might cost more than you expect. IMO it's better not to pessimize a fundamental control construct like this

Reply:
I decided to put these things in as ops. One note: Should they be `aten` or `prim` ops? What's the difference?